pkg/csi_driver/controller.go

```diff
@@ -417,36 +437,99 @@ func (s *controllerServer) ControllerGetCapabilities(ctx context.Context, req *c
 	}, nil
 }
 
-// getRequestCapacity returns the volume size that should be provisioned
-func getRequestCapacity(capRange *csi.CapacityRange) (int64, error) {
-	if capRange == nil {
-		return minVolumeSize, nil
+// getTierFromParams returns the provided tier or default
+func getTierFromParams(params map[string]string) string {
+	if val, ok := params[paramTier]; ok {
+		return val
 	}
 
-	rCap := capRange.GetRequiredBytes()
-	rSet := rCap > 0
-	lCap := capRange.GetLimitBytes()
-	lSet := lCap > 0
+	return defaultTier
+}
+
+func invalidCapacityRange(capRange *csi.CapacityRange, tier string)  error {
+	validRange := provisionableCapacityForTier(tier)
+
+	minRequired := capRange.GetRequiredBytes()
+	minSet := minRequired > 0
+	maxRequested := capRange.GetLimitBytes()
+	maxSet := maxRequested > 0
 
-	if lSet && rSet && lCap < rCap {
-		return 0, fmt.Errorf("limit bytes %v is less than required bytes %v", lCap, rCap)
+
+
+	if maxSet && minSet && maxRequested < minRequired {
+		return fmt.Errorf("limit bytes %v is less than required bytes %v", maxRequested, minRequired)
 	}
 
-	if lSet && lCap < minVolumeSize {
-		return 0, fmt.Errorf("limit bytes %v is less than minimum instance size bytes %v", lCap, minVolumeSize)
+	if minSet{
+		if minRequired > validRange.max {
+			return fmt.Errorf("request bytes %v is more than maximum instance size bytes %v for tier %s", minRequired, validRange.max, tier)
+		}
+
+		if !maxSet && minRequired < validRange.min { 
+			// Avoid surprising users by provisioning more than Requested
+			return fmt.Errorf("limit bytes %v is less than minimum instance size bytes %v for tier %s, but no maximum was specified", minRequired, validRange.min, tier)
+		}
 	}
+	if maxSet {
+		if maxRequested < validRange.min {
+			return fmt.Errorf("limit bytes %v is less than minimum instance size bytes %v for tier %s", maxRequested, validRange.min, tier)
```

Comment:
if we try to surface these message to users, do we want to make them clear by saying in terms of`TiB` instead of bytes?

Reply:
Good point! I followed the existing messages, but the TiB values would be much more legible.